src/main/java/com/diffmin/patch/PatchApplication.java

```diff
@@ -77,6 +77,28 @@ private static void performInsertion(
                                 ((CtExecutable<?>) toBeInserted.getParent()).getThrownTypes());
                 ((CtExecutable<?>) inWhichElement).setThrownTypes(thrownTypesCopy);
                 break;
+            case CONTAINED_TYPE:
+                // Inserting into CtPackage
+                Set<CtType<?>> typesCopy =
+                        new HashSet<>(((CtPackage) toBeInserted.getParent()).getTypes());
+                for (CtType<?> type : typesCopy) {
+                    type.setParent(inWhichElement);
+                }
```

Comment:
This could be combined with line 83 itself and I was trying to do using `stream.map` like so.

```java
new HashSet<>(((CtPackage) toBeInserted.getParent()).getTypes().stream().map(
        (type) -> {
                type.setParent(inWhichElement);
                return type;
         }
));
```

But it kept throwing `Bad return type in lambda expression: CtType<capture of ?> cannot be converted to R`. `R` in my  opinion should be `CtType<?>` only.

Reply:
Also, this loop for setting parent was necessary otherwise it threw a `SpoonException`.
```sh
spoon.SpoonException: The default behavior has changed, a new check has been added! Don't worry, you can disable this check
with one of the following options:
 - by configuring Spoon with getEnvironment().setSelfChecks(true)
 - by removing the node from its previous parent (element.delete())
 - by cloning the node before adding it here (element.clone())
```

I figured that I needed to _detach_ the inserted type from their previous parent so I used `setParent` API. I was also trying to delete the parent using `type.getParent().delete()` but that did not work.

Although the solution was written in the above exception message, I was not able to implement what it was trying to suggest.

Reply:
> But it kept throwing Bad return type in lambda expression: CtType<capture of ?> cannot be converted to R. R in my opinion should be CtType<?> only.

This is probably a type inference failure, you may have to [_capture_ the wildcard](https://docs.oracle.com/javase/tutorial/java/generics/capture.html). Also, this is a bit of a misuse of the `map` method, you really shouldn't cause side effects in it. Streams are functional constructs, and functional constructs should be pure (no side effects). The `Stream.forEach` method is an exception: it's specifically meant for side effects in streams.

> I figured that I needed to detach the inserted type from their previous parent so I used setParent API. I was also trying to delete the parent using type.getParent().delete() but that did not work.

The error message says that you should try `element.delete()`, which translates to `type.delete()`. That removes `type` from its parent, making it safe to set it. What you've done is to remove the parent of type from its parent, which doesn't make sense in the context.

However, you should avoid modifying the new model as you're at this stage still extracting changes from it. See my suggestion below.

Reply:
> What you've done is to remove the parent of type from its parent

I think I have removed the parent of type from the **type** itself. In my opinion, `type.setParent` removes the type from its parent and sets a new parent provided.

> However, you should avoid modifying the new model as you're at this stage still extracting changes from it.

Makes sense. I have one follow-up question. Would this still be a problem if all my patches were generated using `element.clone` instead of directly using the `element`?